Modules/Common/SolanaTraits.php

```diff
@@ -0,0 +1,364 @@
+    function filter_events_by_currency(array $currencies_filter, array $events): array
+    {
+        $filtered = [];
+        $sort_key = 0;
+        foreach ($events as &$event)
+        {
+            if (!in_array($event['currency'], $currencies_filter))
+                continue;
+            if ($this->currency_type === CurrencyType::NFT)
+                if ($event['effect'] !== '1' && $event['effect'] !== '-1')
+                    throw new ModuleError("Invalid effect for NFT token transfer");
+
+            $event['sort_key'] = $sort_key++;
+            $filtered[] = $event;
+        }
+
+        return $filtered;
+    }
+
+    function get_token_supply($currencies): array
+    {
+        $tokens_supply = [];
+        for ($i = 0; $i < count($currencies); $i++)
+        {
+            $data[] = [
+                'method' => 'getTokenSupply',
+                'params' => [$currencies[$i]["id"]],
+                'id' => $i,
+                'jsonrpc' => '2.0',
+            ];
+        }
+
+        $data_chunks = array_chunk($data, 100);
+
+        foreach ($data_chunks as $datai)
+        {
+            $result = requester_single($this->select_node(), params: $datai);
+
+            foreach ($result as $bit)
+            {
+                $tokens_supply[$currencies[$bit['id']]['id']] = $bit['result']['value']['amount'] ?? "0";
+            }
+        }
+        return $tokens_supply;
+    }
+
+    // Returns new program derived address of Metaplex Metadata from mint address
+    function find_metaplex_meta_pda(string $mint): ?string
+    {
+        $fullseeds = 'metadata';
+        $fullseeds .= Base58::base58_nocheck_decode(SolanaAddressPrograms::METAPLEX_ID->value);
+        $fullseeds .= Base58::base58_nocheck_decode($mint);
+
+        return $this->find_program_address($fullseeds, SolanaAddressPrograms::METAPLEX_ID->value);
+    }
+
+// Returns new program derived address of Metaplex Master Edition from mint address
+    function find_program_address($seeds, $program_id): ?string
+    {
+        for ($bump = 255; $bump != 0; $bump--)
+        {
+            $fullseeds_copy = $seeds;
+            $fullseeds_copy .= chr($bump);
+            $fullseeds_copy .= Base58::base58_nocheck_decode($program_id);
+            $fullseeds_copy .= 'ProgramDerivedAddress';
+            $binhash = hash('sha256', $fullseeds_copy, true);
+
+            // Valid program addresses must fall off the ed25519 curve
+            if (!$this->is_on_curve($binhash))
+                return Base58::base58_nocheck_encode($binhash);
+        }
+
+        // Cannot find metaplex pda for this mint
+        return null;
+    }
+
+// Checks that public key is on ed25519 curve. Uses sodium extension.
+    function is_on_curve(string $pk_bytes): bool
+    {
+        try
+        {
+            ParagonIE_Sodium_Compat::crypto_sign_ed25519_pk_to_curve25519($pk_bytes);
+            return true;
+        }
+        catch (Exception $e)
```

Comment:
Can this throw errors? If yes, `Exception` needs to be changed to `Throwable`. Also, `$e` is unnecessary here.

Reply:
Yes and this is actually what we need.